pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/PercentileSmartTDigestAggregationFunction.java

```diff
@@ -0,0 +1,346 @@
+ * exceeds a threshold, the list will be converted into a {@link TDigest}, and approximate result will be returned.
+ *
+ * The function takes an optional third argument for parameters:
+ * - threshold: Threshold of the number of values to trigger the conversion, 100_000 by default. Non-positive value
+ *              means never convert.
+ * - compression: Compression for the converted TDigest, 100 by default.
+ * Example of third argument: 'threshold=10000;compression=50'
+ */
+public class PercentileSmartTDigestAggregationFunction extends BaseSingleInputAggregationFunction<Object, Double> {
+  private static final double DEFAULT_FINAL_RESULT = Double.NEGATIVE_INFINITY;
+
+  private final double _percentile;
+  private final int _threshold;
+  private final int _compression;
+
+  public PercentileSmartTDigestAggregationFunction(List<ExpressionContext> arguments) {
+    super(arguments.get(0));
+    try {
+      _percentile = Double.parseDouble(arguments.get(1).getLiteral());
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+          "Second argument of PERCENTILE_SMART_TDIGEST aggregation function must be a double literal (percentile)");
+    }
+    Preconditions.checkArgument(_percentile >= 0 && _percentile <= 100, "Invalid percentile: %s", _percentile);
+    if (arguments.size() > 2) {
+      Parameters parameters = new Parameters(arguments.get(2).getLiteral());
+      _compression = parameters._compression;
+      _threshold = parameters._threshold;
+    } else {
+      _threshold = Parameters.DEFAULT_THRESHOLD;
+      _compression = Parameters.DEFAULT_COMPRESSION;
+    }
+  }
+
+  public double getPercentile() {
+    return _percentile;
+  }
+
+  public int getThreshold() {
+    return _threshold;
+  }
+
+  public int getCompression() {
+    return _compression;
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.PERCENTILESMARTTDIGEST;
+  }
+
+  @Override
+  public String getColumnName() {
+    return AggregationFunctionType.PERCENTILESMARTTDIGEST.getName() + _percentile + "_" + _expression;
+  }
+
+  @Override
+  public String getResultColumnName() {
+    return AggregationFunctionType.PERCENTILESMARTTDIGEST.getName().toLowerCase() + "(" + _expression + ", "
+        + _percentile + ")";
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    validateValueType(blockValSet);
```

Comment:
Doing this once (upon first call to aggregate) should be enough ?

Reply:
Yes, but because the aggregation function itself is stateless (shared across threads), we cannot add a variable to the function to track if it is the first call. The overhead of this call should be minimal. Once we enforce schema, we should be able to perform all these checks on the broker side